### PR TITLE
Tests for #nodeSubscriptions and subscriptions in UserNodePlayer

### PR DESCRIPTION
**Pull request by cfoust**
**User-Facing Changes**
None.

**Description**
I promised to write a comprehensive set of tests for this complicated transformation with a lot of edge cases, so here we are.

### CODE REVIEW

File: packages/studio-base/src/players/UserNodePlayer/subscriptions.test.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        },
+        {
+          topic: "/test2",
+          preloadType: "partial",
+        },
+      ],
+    ]);
+  });
+
+  it("upgrades to a full subscription from fields", () => {
```
Comment (jtbandes) [inline packages/studio-base/src/players/UserNodePlayer/subscriptions.test.ts:92]: Not sure I fully understand this scenario – why doesn't the output have the `fields` in it anywhere?

Comment (cfoust) [reply packages/studio-base/src/players/UserNodePlayer/subscriptions.test.ts:92]: any inputs to user scripts receive subscriptions to all fields. we also don't allow for subscribing to individual fields of a user node topic

File: packages/studio-base/src/players/UserNodePlayer/index.ts [hunk 2]
```diff
@@ -1042,76 +1042,12 @@ export default class UserNodePlayer implements Player {
...
-              {
-                topic: subscription.topic,
-                preloadType,
-              },
-            ];
-          }
-
-          return [subscription];
-        }
+    const [nodeSubscriptions, resolvedSubscriptions] = simplifySubscriptions(
```
Comment (snosenzo) [inline packages/studio-base/src/players/UserNodePlayer/index.ts:1045]: I think this and the tests might be easier to read if this function was split into two
```
// this actually doesn't need inputsByOutputTopic
this.#nodeSubscriptions = simplifySubscriptions(subscriptions); 

this.#playerSubscriptions.setSubscriptions(
  resolveToRealSubs(subscriptions, state.inputsByOutputTopic)
);
```
I think this function name hides too much and it's hard to see what's being passed where. Splitting the two of these functions up will also make the tests a bit easier to read.

File: packages/studio-base/src/players/UserNodePlayer/subscriptions.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+  const resolvedSubscriptions = R.pipe(
+    R.chain(([subscription, topics]: SubscriberInputs): SubscribePayload[] => {
+      const preloadType = subscription.preloadType ?? "partial";
+
+      // Leave the subscription unmodified if it is not a user script topic
+      if (topics == undefined) {
+        // If this is an input to a user script, we need to upgrade it to a
+        // subscription of all the fields
+        if (neededInputTopics.includes(subscription.topic)) {
```
Comment (snosenzo) [inline packages/studio-base/src/players/UserNodePlayer/subscriptions.ts:70]: actually coming back to this now -- we only need to upgrade this subscription to read all of the fields if the output of the user script is subscribed to

**Comment by cfoust** [comment]
Bump on this!

File: packages/studio-base/src/players/UserNodePlayer/subscriptions.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+  return R.pipe(
+    R.chain(([subscription, topics]: SubscriberInputs): SubscribePayload[] => {
+      const preloadType = subscription.preloadType ?? "partial";
+
+      // Leave the subscription unmodified if it is not a user script topic
+      if (topics == undefined) {
+        // If this is an input to a user script, we need to upgrade it to a
+        // subscription of all the fields
+        if (neededInputTopics.includes(subscription.topic)) {
```
[the remark this answers is not in the document]
Comment (cfoust) [reply packages/studio-base/src/players/UserNodePlayer/subscriptions.ts:70]: I might be misunderstanding what you're saying, but if a topic is an input to a user script we necessarily need a subscription to all of its fields because we can't know what fields the script uses. If that's not what you're referring to, could you clarify what you mean here?

[the remark this answers is not in the document]
Comment (snosenzo) [reply packages/studio-base/src/players/UserNodePlayer/subscriptions.ts:70]: Those input subscriptions to all fields will be produced below where you iterate over the inputs to an output topic and create whole-message subscriptions with the corresponding preloadType. Since those input topic subscriptions are already being produced, mergeSubscriptions should be able to resolve fields subs to whole-message subs as we need.

**Review by snosenzo — APPROVED** [review]
up to you whether you want to include the changes from my branch or not:
https://github.com/foxglove/studio/compare/caleb/fg-4952-tests-for-nodesubscriptions-and-subscriptions-in...no-coalesce-input-subs?expand=1
